### PR TITLE
ci: add formatting check with each PR

### PR DESCRIPTION
**Pull request by antonbaliasnikov**
# Code Review Checklist

## Purpose

Adds diff formatting check with each PR.

## Ticket Number

Fixes [CPR-1054](https://linear.app/matterlabs/issue/CPR-1054/add-clang-format-check-to-the-ci-run)

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/utils/git/code-format-helper-era-llvm.py
Comment (FlashSheridan) [inline llvm/utils/git/code-format-helper-era-llvm.py:49]: _Not_ our problem, but Pylint has a number of stylistic objections.  For this line, it says:

> test for object identity should be 'is not'

File: llvm/utils/git/code-format-helper-era-llvm.py
Comment (FlashSheridan) [inline llvm/utils/git/code-format-helper-era-llvm.py:351]: Also not our problem:  Grammatical error.
(If we contribute something else upstream, we might include Pylint warnings fixes.  Not worth an independent complaint, lest we look like airdrop farmers.)

File: llvm/utils/git/code-format-helper-era-llvm.py
Comment (FlashSheridan) [inline llvm/utils/git/code-format-helper-era-llvm.py:330]: Upstream misspelling:  “authentiation”

File: llvm/utils/git/code-format-helper-era-llvm.py
Comment (FlashSheridan) [inline llvm/utils/git/code-format-helper-era-llvm.py:1]: This presumably doesn’t affect the GitHub Action, but trying to run this script on my Mac led to a world of version pain.  Running the script initially gave an error “No module named 'github'”.  Installing the module (`pip3 install github`) inevitably fails with “aiohttp/_websocket.c:198:12: fatal error: 'longintrepr.h' file not found”.

The underlying problem may be with https://github.com/aio-libs/aiohttp/issues/6600, which led to the grim suggestion:

 https://github.com/cython/cython/pull/4428#issuecomment-1682593857
>> I can't update cython version as its not compatible with coqui-tts

> Then you're stuck and there is no possible solution that lets you use Python 3.11.
> I suggest you use Python 3.10 or earlier.


Something had indeed switched my Mac’s Python3 from Apple’s /usr/bin/python3 (3.9.6) to /opt/homebrew/bin/python3 (3.11.6), which is inherently dangerous.  But switching back (mv /opt/homebrew/bin/python3 /opt/homebrew/bin/python3_Homebrew) and trying to (re-)install the github module failed the same way.


I’m still concerned about formatting complaints on existing code, but I can probably see those without the wrapper script, which after all doesn’t really need to run on my Mac, though it would be convenient for sufficiently diligent developers.  (Experience with clang-tidy suggests that that is not a major concern.)

File: .github/workflows/pr-code-format.yml
Comment (FlashSheridan) [inline .github/workflows/pr-code-format.yml:1]: I would like to see deeper testing on C++ formatting before we merge; I didn’t see any C++ files getting checked (e.g. [runs/8254177705/job/22577714626#step:9:19](https://github.com/matter-labs/compiler-llvm/actions/runs/8254177705/job/22577714626#step:9:19) has a Python, Yaml, and text file, and  [runs/8254177705/job/22577714626#step:9:1](https://github.com/matter-labs/compiler-llvm/actions/runs/8254177705/job/22577714626#step:9:1)) has only a text file changed.

File: llvm/utils/git/code-format-helper-era-llvm.py
Comment (FlashSheridan) [inline llvm/utils/git/code-format-helper-era-llvm.py:186]: For those trying this at home, I think [job/22577714626#step:6:160](https://github.com/matter-labs/compiler-llvm/actions/runs/8254177705/job/22577714626#step:6:160) means that `apt` is installing a binary `clang-format` (17.0.6), though llvm/clang/tools/clang-format/git-clang-format’s usage says it‘s a git command (“usage = ('git clang-format…”.)

The Mac equivalent is presumably “brew install clang-format” (also 17.0.6).  Once that was in my path, I could execute:

``` clang/tools/clang-format/git-clang-format origin/dev-17 ```

This seemed to change about 4510 files, but only 9 matching “Era”.  An example change in our hierarchy:


``` +++ b/llvm/lib/Target/EraVM/EraVMAsmPrinter.cpp
@@ -165,8 +165,7 @@ void EraVMAsmPrinter::emitGlobalConstant(const DataLayout &DL,
     return;
   }
 
-  if (const auto *CDS =
-          dyn_cast<ConstantDataSequential>(CV)) {
+  if (const auto *CDS = dyn_cast<ConstantDataSequential>(CV)) {
     size_t elem_size = CDS->getNumElements();
```

Comment (FlashSheridan) [reply llvm/utils/git/code-format-helper-era-llvm.py:186]: On changes more specific to us, last  week,

```clang/tools/clang-format/git-clang-format cbee545ae03584a5a2509520ec1541ff99bac563^```
changed 11 files:

    llvm/include/llvm/ADT/Triple.h
    llvm/include/llvm/MC/MCDecoderOps.h
    llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp
    llvm/lib/MC/MCExpr.cpp
    llvm/lib/Support/Triple.cpp
    llvm/lib/Target/EVM/EVMISelLowering.cpp
    llvm/lib/Target/EVM/EVMMachineFunctionInfo.h
    llvm/lib/Target/EVM/EVMSingleUseExpression.cpp
    llvm/lib/Target/EVM/EVMTargetMachine.cpp
    llvm/lib/Target/EVM/MCTargetDesc/EVMMCExpr.h
    llvm/utils/TableGen/DecoderEmitter.cpp

Here’s one of the diffs:

```
+++ b/llvm/lib/Target/EVM/EVMISelLowering.cpp
@@ -436,8 +436,8 @@ SDValue EVMTargetLowering::LowerCall(CallLoweringInfo &CLI,
                          getPointerTy(DAG.getDataLayout()), Callee);
   } else if (Callee->getOpcode() == ISD::ExternalSymbol) {
     auto *ES = cast<ExternalSymbolSDNode>(Callee);
-    Callee = DAG.getTargetExternalSymbol(ES->getSymbol(),
-                                         Callee.getValueType());
+    Callee =
+        DAG.getTargetExternalSymbol(ES->getSymbol(), Callee.getValueType());
     Callee = DAG.getNode(EVMISD::TARGET_ADDR_WRAPPER, DL,

```

**Review by FlashSheridan — COMMENTED** [review]
There’s an engineering policy question for Dmitry and his team:  Are they expected to format (and check) locally code which they’ve touched before pushing, and/or are they willing to accept a certain amount of inconvenience from GitHub Action failures?  The local testing I’ve been able to do (see previous comments) suggests that this won’t be an inordinate amount, but may be non-trivial.
    A similar issue for clang-tidy had more GitHub Action failures than I’d hoped for, but didn’t elicit any objections that I’m aware of.

**Comment by akiramenai** [comment]
@FlashSheridan if I got the question right, we are ok to have a failure because of clang-format on CI. It's much better than merge an unformatted patch and then have irrelevant formatting changes committed with something else. We had a lot of that in the past, so I hope this PR will prevent such things in future.

**Comment by FlashSheridan** [comment]
> @FlashSheridan if I got the question right, we are ok to have a failure because of clang-format on CI. It's much better than …

Agreed, and anyway that‘s an engineering policy decision which QA would be wise to stay out of.  I was hoping for a third way, though, where an engineer could be sure before committing that his patch was properly formatted.  (E.g. with a Git hook exactly matching CI, because it used the same script.)  I think my technique above improves the odds, but I don’t know whether it properly handles configuration.

(Sorry for the slow response — taking PTO for more revisions for Logique et Analyse.)

File: .github/workflows/pr-code-format.yml
Comment (antonbaliasnikov) [reply .github/workflows/pr-code-format.yml:1]: @FlashSheridan , here is an example of a C++ formatting check that I made and then reverted changes:

https://github.com/matter-labs/compiler-llvm/actions/runs/8253921783/job/22576874677

File: llvm/utils/git/code-format-helper-era-llvm.py
Comment (akiramenai) [reply llvm/utils/git/code-format-helper-era-llvm.py:186]: It's expected as we didn't have `clang-format` in CI. Anyway, I'm reformatting on top of LLVM 17.

**Review by akiramenai — APPROVED** [review]

File: .github/workflows/pr-code-format.yml
Comment (FlashSheridan) [reply .github/workflows/pr-code-format.yml:1]: Excellent, thanks.  (It wasn’t obvious to me from the web interface what the relevant source change was, but the [log](https://github.com/matter-labs/compiler-llvm/actions/runs/8253921783/job/22576874677#step:2:180) says a21425e, which looks good.)

**Review by hedgar2017 — APPROVED** [review]
LGTM!
Let's just prevent false-positives if possible, sir.

**Comment by akiramenai** [comment]
@antonbaliasnikov what prevents us from squashing and merging?

**Comment by antonbaliasnikov** [comment]
> @antonbaliasnikov what prevents us from squashing and merging?

@akiramenai , nothing. I was just waiting for the final confirmation that we're ok with the current implementation and behavior after reviews. I'm rebasing and merging then.

**Comment by akiramenai** [comment]
You got 2 LGTMs. If you think that we might overlooked something particular, let's discuss the concern. Otherwise, let's get it merged.

**Pull request merged**
